### ISSUE TITLE
Duplicate scenarios running in parallel

### ISSUE THREAD

**Issue by sgupta304** [prashant-ramcharan/courgette-jvm#54]
Hi Prashant,

So the situation I have is somewhat unique. I'm trying to run each scenario in parallel on different browsers. Pictured below:
.........**Chrome**...**Firefox**
**Sc1**...Sc1_Ch...Sc1_FF
**Sc2**...Sc2_Ch...Sc2_FF

I'm able to successfully invoke the courgette runner with the scenarios, however, if I run 2 scenarios on 2 browsers. The number of actual scenarios that are supposed to run are 4, however 16 threads get kicked off. And I have duplicate scenarios running in parallel. If I run 2 scenarios on a single browser then 4 threads get kicked off, when in fact it is supposed to be only 2. The reports shows as seen here: 
![image](https://user-images.githubusercontent.com/9002815/40401908-8863e1cc-5e0e-11e8-92b9-796cbd5c4070.png). 

I'm sorry if my problem is a little confusing. Please let me know if you need more information from me. 

Your runner invokes the correct number of threads the first time. It just invokes it by a factor of the number of scenarios. So if I am running 2 scenarios on 2 browsers. It kicks off 4 threads 4 times.

**Comment by prashant-ramcharan** [comment]
I don’t understand the issue here.

You have 4 scenarios which run as expected in parallel. The Courgette-JVM report shows that only 4 scenarios ran.

Please share your project or re-create the issue with the Courgette-JVM example project.

Thanks

**Comment by sgupta304** [comment]
Here you go https://github.com/sgupta304/wdraTEz3ieJAbC9Zp_rep. When you build the project you will see the build pattern is gradle->testng.xml->TestSetupRunner->TestSuiteRunner. I'm trying to run the same scenario on two different browsers in parallel.

**Comment by prashant-ramcharan** [comment]
I took a look and its running as expected.

When I run **TestSuiteRunner**, it runs 2 scenarios with each scenario opening a single browser (chrome)

It only runs 2 scenarios -- which is expected

-----

What you trying to do is run a **single cucumber scenario** on 2 browsers. This cannot happen because your cucumber test only uses one browser at a time.

What you need to do is **run both** suite runners (below) at the same time.

Create a **ChromeTestSuiteRunner**.

````java
@Test
@CourgetteOptions(
   threads = 10,
   runLevel = CourgetteRunLevel.SCENARIO,
   rerunFailedScenarios = true,
   showTestOutput = true,
   reportTargetDir = "build/chrome",
   cucumberOptions = @CucumberOptions(
		features = "src/test/resources/features",
		glue = "Application.StepDefinitions",
		tags = {"@healthCheck"},
		plugin = {
			"progress",
			"json:build/chrome/cucumber-report/cucumber.json",
			"html:build/chrome/cucumber-report/cucumber.html"},
			strict = true
				))
public class ChromeTestSuiteRunner extends TestNGCourgette {
}
````

Create a **FirefoxTestSuiteRunner**.

````java
@Test
@CourgetteOptions(
   threads = 10,
   runLevel = CourgetteRunLevel.SCENARIO,
   rerunFailedScenarios = true,
   showTestOutput = true,
   reportTargetDir = "build/firefox",
   cucumberOptions = @CucumberOptions(
		features = "src/test/resources/features",
		glue = "Application.StepDefinitions",
		tags = {"@healthCheck"},
		plugin = {
			"progress",
			"json:build/firefox/cucumber-report/cucumber.json",
			"html:build/firefox/cucumber-report/cucumber.html"},
			strict = true
				))
public class FirefoxTestSuiteRunner extends TestNGCourgette {
}
````

You need to pass a System property to decide the browser at runtime. For example, in your driver setup read from -**DbrowserType** to decide the browser to use.

**Comment by prashant-ramcharan** [comment]
There's no logical way for a **single test runner** to know that you want run a test that currently opens a single browser to then open multiple browsers...

Unless you specifically code this logic in your cucumber tests!

**Comment by sgupta304** [comment]
Hi Prashant, 

I've updated my code to be able to handle different browsers using a single courgette runner. However, it is still kicking multiple instances of chrome and firefox. More than needed. For how the current scenario is setup, it should only kick off 4 scenarios, but it is kicking off 16-17

**Comment by prashant-ramcharan** [comment]
Please provide the run command you use to run your tests.

Are you simple calling:

gradle test

**Comment by sgupta304** [comment]
Yes. Just Gradle test

**Comment by prashant-ramcharan** [comment]
Thanks. 

But what system properties do you pass to gradle test?

**Comment by sgupta304** [comment]
Nothing else is needed for now. Things are hardcoded right now as this is still in testing phase. I've hard coded the browser specifications in the config.properties

**Comment by prashant-ramcharan** [comment]
java.lang.ArrayIndexOutOfBoundsException: 3
	at Application.StepDefinitions.BaseStepDefinitions.getDriverBrowser(BaseStepDefinitions.java:60)
	at Application.StepDefinitions.BaseStepDefinitions.setupDriver(BaseStepDefinitions.java:51)
	at cucumber.runtime.java8.Java8HookDefinition.lambda$execute$0(Java8HookDefinition.java:49)
	at cucumber.runtime.Timeout.timeout(Timeout.java:16)
	at cucumber.runtime.java8.Java8HookDefinition.execute(Java8HookDefinition.java:47)
	at cucumber.runtime.HookDefinitionMatch.runStep(HookDefinitionMatch.java:17)
	at cucumber.runner.UnskipableStep.executeStep(UnskipableStep.java:22)
	at cucumber.api.TestStep.run(TestStep.java:83)
	at cucumber.api.TestCase.run(TestCase.java:58)
	at cucumber.runner.Runner.runPickle(Runner.java:80)
	at cucumber.runtime.Runtime.runFeature(Runtime.java:119)
	at cucumber.runtime.Runtime.run(Runtime.java:104)
	at cucumber.api.cli.Main.run(Main.java:36)
	at cucumber.api.cli.Main.main(Main.java:18)

**Comment by sgupta304** [comment]
Are you running on a windows or a mac?

**Comment by prashant-ramcharan** [comment]
Another thing... theres no need to use **TestNGCucumberRunner**

TestNGCucumberRunner is the cucumber runner for testng and is likely the reason why your tests are being run more than once.

Your tests are being run with TestNGCucumberRunner and Courgette.

Remove the following:

- TestNGCucumberRunner
-  systemProperty "cucumber.options", System.getProperty("cucumber.options")

**Comment by sgupta304** [comment]
I'm using that to get the current number of scenarios. Is there another way for me to get the number of scenarios based on the tags?

**Comment by sgupta304** [comment]
I'm not invoking the TestNGCucumberRunner, I thought? The only way that would get invoked would be if I did TestNGCucumberRunner.runScenarios(). Right?

**Comment by sgupta304** [comment]
**Why do you need to get the number of scenarios? Cucumber / Courgette already does this for you using its Runtime**

I'm using the number of scenarios and multiplying it with the number of browsers coming from the config.properties to get a new set of feature files that are stored in the /tmp folder. That is where your courgette runner is picking them up from. Each scenario gets multiplied into the number of browsers. Although the scenario steps stay the same, the name of the feature and the scenario name changes. This was a workaround for getting a single scenario to run on two different browsers at the same time.

![image](https://user-images.githubusercontent.com/9002815/40440754-8a9274b0-5e84-11e8-919a-abe9f5fb35b9.png)
The browser type is coming from the name of the feature.

**Comment by sgupta304** [comment]
**I think this project setup and the use of Courgette is incorrect.
I will add an example project that demonstrates a scenario that run on both browsers.**

Thank you so much @prashant-ramcharan

**Comment by sgupta304** [comment]
Would your solution be scalable? Meaning if tomorrow I wanted to added different types of browsers to a single scenario, it would still work?

**Comment by sgupta304** [comment]
Prashant, I commented out the testNgCucumberRunner and the system property to read cucumber.options as well.

`package Framework.Runner;

import Framework.Configuration.EnvironmentOption;
import Framework.Utilities.ConfigFileReader;
import cucumber.api.CucumberOptions;
import cucumber.api.testng.PickleEventWrapper;
import cucumber.api.testng.TestNGCucumberRunner;
import gherkin.events.PickleEvent;
import org.testng.TestNG;
import org.testng.annotations.AfterTest;
import org.testng.annotations.BeforeTest;
import org.testng.annotations.Test;
import org.testng.xml.XmlClass;
import org.testng.xml.XmlSuite;
import org.testng.xml.XmlTest;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

@CucumberOptions(
  tags = {"@healthCheck"},
  features = "src/test/resources"
)
public class TestSetupRunner {
//  private TestNGCucumberRunner testNGCucumberRunner;
//  private Object[][] scenarios;
  private TestNG testNg;
  private XmlSuite suite;
  private ArrayList<XmlSuite> suites;
  private String PRESENT_WORKING_DIRECTORY = System.getProperty("user.dir");

  @BeforeTest
  public void setup() {
    testNg = new TestNG();
    suite = new XmlSuite();
    suite.setName("Run Suite");
    testNg.setJUnit(false);
    testNg.setUseDefaultListeners(false);
    testNg.setOutputDirectory("/local/testNg");
//    testNGCucumberRunner = new TestNGCucumberRunner(this.getClass());
    writeFeaturesToBuildDirectory();
  }

  @Test
  public void test() {
    ArrayList<XmlClass> classes = new ArrayList<>();
    XmlTest test = new XmlTest(suite);
    test.setName("Run Test");
    classes.add(new XmlClass("Framework.Runner.TestSuiteRunner"));
    test.setXmlClasses(classes);
  }

  @AfterTest
  public void afterTest() {
    suites = new ArrayList<>();
    suites.add(suite);
    testNg.setXmlSuites(suites);
//    testNGCucumberRunner.finish();
    testNg.run();
  }

  private List<EnvironmentOption> getEnvOption() {
  	String configFile = "/src/test/resources/Properties/config";
    List<EnvironmentOption> options = new ArrayList<>();
    ConfigFileReader cReader = new ConfigFileReader(PRESENT_WORKING_DIRECTORY+configFile);
    String envNum = cReader.getPropFile().getProperty("RunEnvrionmentOption");
    String[] envCombo = cReader.getPropFile().getProperty("EnvironmentOption_" + envNum).split(",");
    for (String combo : envCombo) {
      String[] option = cReader.getPropFile().getProperty(combo).split("\\|");
      options.add(new EnvironmentOption(option[1], option[2], option[0]));
    }
    return options;
  }

//  private void writeFeaturesToBuildDirectory() {
//    scenarios = testNGCucumberRunner.provideScenarios();
//    List<EnvironmentOption> options = getEnvOption();
//    for (int i = 0; i <= scenarios.length - 1; i++) {
//      PickleEvent scenario = ((PickleEventWrapper) scenarios[i][0]).getPickleEvent();
//      for (EnvironmentOption option : options) {
//        String fName =
//            String.format(
//                "%s_%s_%s_%s.feature",
//                scenario.uri.split("/")[4].replace(".feature", ""),
//                option.getPlatform(),
//                option.getBrowser(),
//                option.getBrowserVersion());
//        writeScenarioToFile(
//            readAndModifyScenario(scenario, option), "/tmp/" + fName.replaceAll(" ", "_"));
//      }
//    }
//  }

  private void writeFeaturesToBuildDirectory() {
    List<EnvironmentOption> options = getEnvOption();
    for (EnvironmentOption option : options) {
      String fName =
              String.format(
                      "OpenYahoo_%s_%s_%s.feature",
                      option.getPlatform(),
                      option.getBrowser(),
                      option.getBrowserVersion());
      writeScenarioToFile(
              readAndModifyScenario(option), "/tmp/" + fName.replaceAll(" ", "_"));
    }
  }

  private String readAndModifyScenario(EnvironmentOption option) {
    StringBuilder sb = new StringBuilder();
    try (BufferedReader reader =
                 Files.newBufferedReader(Paths.get("/src/test/resources/features/OpenYahoo.feature"))) {
      String line = null;
      while ((line = reader.readLine()) != null) {
        if (line.contains("Scenario")) {
          sb.append("Scenario: ")
                  .append("Validate Yahoo.com loads properly")
                  .append(" on ")
                  .append(
                          String.format(
                                  "%s %s %s",
                                  option.getBrowser(), option.getBrowserVersion(), option.getPlatform()))
                  .append("\n");
        } else {
          sb.append(line).append("\n");
        }
      }
    } catch (IOException x) {
      System.err.format("IOException: %s%n", x);
    }
    return sb.toString();
  }

//  private String readAndModifyScenario(PickleEvent scenario, EnvironmentOption option) {
//    StringBuilder sb = new StringBuilder();
//    try (BufferedReader reader =
//        Files.newBufferedReader(Paths.get(PRESENT_WORKING_DIRECTORY + "/" + scenario.uri))) {
//      String line = null;
//      while ((line = reader.readLine()) != null) {
//        if (line.contains("Scenario")) {
//          sb.append("Scenario: ")
//              .append(scenario.pickle.getName())
//              .append(" on ")
//              .append(
//                  String.format(
//                      "%s %s %s",
//                      option.getBrowser(), option.getBrowserVersion(), option.getPlatform()))
//              .append("\n");
//        } else {
//          sb.append(line).append("\n");
//        }
//      }
//    } catch (IOException x) {
//      System.err.format("IOException: %s%n", x);
//    }
//    return sb.toString();
//  }

  private void writeScenarioToFile(String content, String fileName) {
    try (BufferedWriter writer = Files.newBufferedWriter(Paths.get(fileName))) {
      writer.write(content);
    } catch (IOException e) {
      System.err.format("IOException: %s%n", e);
    }
  }
}
`
Now the only runner I have is the courgette runner. However, I still the single scenario of two browsers getting kicked off 4 times :(. Twice on chrome and twice on Firefox

**Comment by prashant-ramcharan** [comment]
_Would your solution be scalable? Meaning if tomorrow I wanted to added different types of browsers to a single scenario, it would still work?_

A more valid question is: "Would your framework scale to handle different browsers?"

**Remember,  Courgettes only responsibility is to run Cucumber test in parallel**.

If you can successfully run a **single Cucumber scenario** against 2 browsers using the standard cucumber runner, then Courgette will be able to run it in parallel.

This library is not meant to scale your browser testing! Its meant to scale your Cucumber test execution.

This issue seems to be directly related to your current framework architecture. Courgette is not being used correctly with your setup.

You cannot run a single scenario against 2 browsers without running two instances of the runner. Its just not possible!

Can you replicate this issue you have with this project: [Courgette-JVM example](https://github.com/prashant-ramcharan/courgette-jvm-example)

**Comment by sgupta304** [comment]
Prashant, I cloned your repo I'm looking  at it right now. In the meanwhile, I updated the example with my code and saved it under my repo, please check here https://github.com/sgupta304/courgette-example

**Comment by prashant-ramcharan** [comment]
Why is your Cucumber options pointing to a tmp directory?

````java
                features = "/tmp",
````

**Comment by sgupta304** [comment]
That is where I write the new feature files. I rewrite the feature files with the browser information in the feature name. I duplicate each feature that is found in `src/resources/feature` into the `/tmp` folder. Each feature found gets multiplied by the number/type of browser I need to run the scenario on. So essentially:

`scenario.feature x chrome = scenario_chrome.feature
scenario.feature x firefox = scenario_firefox.feature`

Both of these will go in the `/tmp` folder where they would be picked up by the courgette runner as a normal folder containing the features. Courgette should not know that the features are copies of the original since the feature and scenario names have been changed.

**Comment by sgupta304** [comment]
are you running this on a mac?

**Comment by prashant-ramcharan** [comment]
This is very confusing and unfortunately out of scope for further investigation.

Out of curiosity, does the current approach work with the normal TestNG cucumber runner?

**Comment by prashant-ramcharan** [comment]
Yes running on Mac

**Comment by sgupta304** [comment]
Your solution did work. I might go forward with your approach as it is less confusing. Thanks for all your help and patience @prashant-ramcharan. I really appreciate your help.

One quick question, is it possible to combine the reporting from all the runs in the end?

**Comment by sgupta304** [comment]
I will keep that in mind!

**Comment by sgupta304** [comment]
Just FYI, here is what I was trying to accomplish. Not sure if this makes it any clearer. 
![20180523_163925](https://user-images.githubusercontent.com/9002815/40452675-470fb242-5ea8-11e8-9674-cb43f9b8374c.jpg)

**Comment by prashant-ramcharan** [comment]
In that case, how sure are you that in the /tmp directory there isn't duplicate feature files? This could very well mean Courgette / Cucumber picks up more than one feature with the same name and therefore runs it more than once. 

If so, then it would explain this conundrum.

**Comment by sgupta304** [comment]
I deleted everything in the `/tmp` before every execution. So it was starting empty every time

**Comment by prashant-ramcharan** [comment]
It could but the process that rewrites the cucumber feature files to /tmp directory needs to happen outside of test execution.

 The problem is, the task which writes to /tmp is triggered from inside your test which is likely the cause. Each thread would start rewriting to the /tmp which would loop over and over.

Move the rewriting of feature files to outside your test, perhaps add it as a gradle task which runs before the test once.

Simply try with disabling the step that writes to /tmp and copy and paste your features in resources/features...

**Comment by sgupta304** [comment]
Hmm...I had a similar hunch. I'll try moving the rewrite to a python script that run before! Thanks again for the help Prashant!
